File: 1538/guess-the-majority-in-a-hidden-array.java
/**
 * // This is the ArrayReader's API interface.
 * // You should not implement it, or speculate about its implementation
 * interface ArrayReader {
 *   public:
 *     // Compares 4 different elements in the array
 *     // return 4 if the values of the 4 elements are the same (0 or 1).
 *     // return 2 if three elements have a value equal to 0 and one element has value equal to 1 or vice versa.
 *     // return 0 : if two element have a value equal to 0 and two elements have a value equal to 1.
 *     public int query(int a, int b, int c, int d);
 *
 *     // Returns the length of the array
 *     public int length();
 * };
 */

class Solution {
    public int guessMajority(ArrayReader reader) {
        int N = reader.length();
        int [] query = new int[N]; //query對照表，用來比對與a[0]是否相同
        boolean [] same = new boolean[N]; //確認是否與a[0]相同

        //能不能從 query 把每個答案找出來? query數字的意思是 2群差多少
        //所以若有4 則代表大家都同一國，很好
        //但若都沒有4就慘了
        //===以下的思考都沒有用，請省略
        //假設 a[0] = 0; query(0,1,2,3) 
        //規則：map如果都是0,要看N是奇數(奇數筆) or 偶數(-1無法判斷)
        //規則：只要有4，就有辦法推算出全部的值，因為可用3個去掃出另外1個的值
        //規則：若有2，就是3打1，有辦法推出4的組合嗎？
        //===以上的思考都沒有，請省略
        query[0] = reader.query(0,1,2,3); 
        same[0] = true;
        int vote0=1;//本來0就要投自己1票
        int other=0;
        //System.out.print("[0]:" + query[0] + " [4]:");
        for(int i=4; i<N; i++){
            query[i] = reader.query(1,2,3,i);
            if(query[0] == query[i]){
                same[i] = true; //與 a[0]相同   
                vote0++;
            }else other = i;
            //System.out.print(query[i] + " "); //印出的對照表
        }
        //太少的話，不好查看。如果n>=6，就可有3個去夾擊3人
        //n:5 不好處理，但還是可查出來，因使用2次查詢來夾擊
        if(N>=7){
            for(int i=1; i<=3; i++){ //把不知道的2,3,4都找出來
                int Q = reader.query(0,i,5,6); //query的結果
                same[i] = testSame(true, same[5], same[6], Q);
                if(same[i] == true) vote0++;
                else other = i;
            }
        } else if(N==5 || N==6){
            int Q10 = reader.query(0,2,3,4);
            int Q11 = reader.query(1,2,3,4);
            if(Q10==Q11) same[1]=true;
            else same[1] = false;
            int Q20 = reader.query(0,1,3,4);
            int Q22 = reader.query(1,2,3,4);
            if(Q20==Q22) same[2]=true;
            else same[2] = false;
            if(N==6){
                int Q30 = reader.query(0,1,4,5);
                int Q33 = reader.query(1,3,4,5);
                if(Q30==Q33) same[3]=true;
                else same[3] = false;
            }
        }

        int voteSame=0, voteOther=0, indexSame=0, indexOther=0;
        for(int i=0; i<N; i++){
System.out.print(same[i] + " ");
            if(same[i]==true){
                voteSame++;
                indexSame = i;
            }else {
                voteOther++;
                indexOther = i;
            }
        }
System.out.println("voteSame:" + voteSame);
        if(voteSame>voteOther) return indexSame;
        else if(voteOther>voteSame) return indexOther;
        else return -1;
//System.out.println(vote0);
//        if(vote0>N/2) return 0; //投a[0]一票的人比較多
//        else if(vote0<N/2) return other; //投另外其他的人
//        else return -1;
    }
    boolean testSame(boolean a, boolean c, boolean d, int query) {
        if(c==true && d==true && query==2) return false;
        else if(c==true && d==true && query==4) return true;
        else if( (c^d) && query==0) return false;
        else if( (c^d) && query==2) return true;
        else if(c==false && d==false && query==2) return false;
        else return true ;//if(c==false && d==false && query==0) return true;
        // true ?? true true 可得到 2 or 4
        // true ?? true false 可得到 0 or 2
        // true ?? false false 可得到 2 or 0
    }
}//case 6/135: [0,0,0,0,1]
//case 7/135: [0,0,1,1,0,1]
//case 112/135: [1,1,0,1,1,0,1,0,0,0]

File: 1538/guess-the-majority-in-a-hidden-array.cpp
class Solution {
public:
    int guessMajority(ArrayReader &reader) {
        // Idea: Keep/Maintain 3 Points of Contact, only 1 Point Move
        // The Idea is used in "Ladder Safety" and "Climbing Safety"
        // - "3 Points Of Contact Rule Of Ladder Safety"
        // - "Maintain Three Points of Contact for Climbing Safety"

        int N = reader.length(); //array lengh of a[i]
        bool same[N]; 
        // same[i] means "a[0] and a[i] are the same"

        // The Idea is to test all a[i] if they are the same as a[0]
        // if(a[0] == a[i]) same[i] = true;
        // else same[i] = false;

        // However, you can't find the Invisible a[i].
        // Therefore, you can use query() to do the similar thing.

        // ex. int Q0 = reader.query(0, 2,3,4);
        //     int Q1 = reader.query(1, 2,3,4) ;
        // if(Q0 == Q1) same[1] = true; //a[0] and a[1] are the same
        // else same[1] = false;    //a[0] and a[1] are not the same

        // In short, you can write: same[1] = (Q0 == Q1);
        
        // Careful: query(a,b,c,d) where a<b<c<d 
        // So you can't use query(0,0,0,0) vs. query(0,0,0,i) 

        //=======
        //Part 1: The first 5 elements (combination -- not easy)

        // For the first 5 elements (0, 1,2,3,4)
        same[0] = true; //of course, a[0] equals a[0], the same

        // Find same[1], same[2], same[3], same[4]
        // 1,2,3,4 are your choices
        // if you want to find 1, keep/maintain 2,3,4
        // if you want to find 2, keep/maintain 1,3,4
        // if you want to find 3, keep/maintain 1,2,4
        // (same[4] can easily be found in Part 2)

        int Q10 = reader.query(0,  2,3,4);
        int Q11 = reader.query(  1,2,3,4);
        same[1] = (Q10==Q11); //maintain 2,3,4 => find 1

        int Q20 = reader.query(0,1,  3,4);
        int Q22 = reader.query(  1,2,3,4);
        same[2] = (Q20==Q22); //maintain 1,3,4 => find 2

        int Q30 = reader.query(0,1,2,  4); 
        int Q33 = reader.query(  1,2,3,4);
        same[3] = (Q30==Q33); //maintain 1,2,4 => find 3

        //=======
        //Part 2: The others (use for loop -- easy)
        // Idea: Keep/Maintain 1,2,3, find i: 4,5,6...
        // int Q0 = reader.query(0, 1,2,3);
        // same[4] =  (reader.query(1,2,3, 4) == Q0); //4 vs. 0
        // same[5] =  (reader.query(1,2,3, 5) == Q0); //5 vs. 0
        // ...
        // same[i] =  (reader.query(1,2,3, i) == Q0); //i vs. 0

        int Q0 = reader.query(0, 1,2,3); 
        for(int i=4; i<N; i++){
            same[i] = (reader.query(1,2,3, i) == Q0);
        }

        //Finally, vote/count same[i] 
        int voteSame=0, voteOther=0;
        int indexSame=0, indexOther=0; //index will be return latter
        for(int i=0; i<N; i++){
            if(same[i]==true){
                voteSame++; //a[i] & a[0] are the same. vote Same
                //indexSame = i; //not necessary. you can return 0 directly
            }else {
                voteOther++; //a[i] & a[0] are not the same, vote Other
                indexOther = i; //remember the index of the other
            }
        }
        if(voteSame>voteOther) return 0; //a[0] win, return 0
        else if(voteOther>voteSame) return indexOther;//the others win
        else return -1; //if the votes of two groups equal, return -1
    }
};
